ArduCopter/mode_systemid.cpp

```diff
@@ -0,0 +1,361 @@
+    }
+
+    if(log_subsample <= 0){
+        log_data();
+        if(copter.should_log(MASK_LOG_ATTITUDE_FAST) && copter.should_log(MASK_LOG_ATTITUDE_MED)) {
+            log_subsample = 1;
+        } else if(copter.should_log(MASK_LOG_ATTITUDE_FAST)){
+            log_subsample = 2;
+        } else if(copter.should_log(MASK_LOG_ATTITUDE_MED)){
+            log_subsample = 4;
+        } else {
+            log_subsample = 8;
+        }
+    }
+    log_subsample -= 1;
+
+    log_data();
+}
+
+// init_test - initialises the test
+void ModeSystemId::log_data()
+{
+    int8_t index = copter.EKF2.getPrimaryCoreIMUIndex();
+    if(index < 0){
+        index = 0;
+    }
+
+    Vector3f delta_angle;
+    copter.ins.get_delta_angle(index, delta_angle);
+    float delta_angle_dt = copter.ins.get_delta_angle_dt(index);
+
+    Vector3f delta_velocity;
+    copter.ins.get_delta_velocity(index, delta_velocity);
+    float delta_velocity_dt = copter.ins.get_delta_velocity_dt(index);
+
+    if(is_positive(delta_angle_dt) && is_positive(delta_velocity_dt)){
+        AP::logger().Write(
+            "SIDI",
+            "TimeUS,Time,Targ,F,Gx,Gy,Gz,Ax,Ay,Az",
+            "ss-zkkkooo",
+            "F---------",
+            "Qfffffffff",
+            AP_HAL::micros64(),
+            (double)waveformTime,
+            (double)waveformSample,
+            (double)waveformFreqRads / (2 * M_PI),
+            (double)(degrees(delta_angle.x / delta_angle_dt)),
+            (double)(degrees(delta_angle.y / delta_angle_dt)),
+            (double)(degrees(delta_angle.z / delta_angle_dt)),
+            (double)(delta_velocity.x / delta_velocity_dt),
+            (double)(delta_velocity.y / delta_velocity_dt),
+            (double)(delta_velocity.z / delta_velocity_dt));
+    }
+    // Full rate logging of attitude, rate and pid loops
+    copter.Log_Write_Attitude();
+}
+
+// init_test - initialises the test
+float ModeSystemId::waveform(float time)
+{
+    float wMin = 2 * M_PI * fStart;
+    float wMax = 2 * M_PI * fStop;
+
+    float window;
+    float output;
+
+    float B = logf(wMax / wMin);
+
+    if(time <= 0.0f) {
+        window = 0.0f;
+    } else if (time <= tFadeIn) {
+        window = 0.5 - 0.5 * cosf(M_PI * time / tFadeIn);
+    } else if (time <= tRec - tFadeOut) {
+        window = 1.0;
+    } else if (time <= tRec) {
+        window = 0.5 - 0.5 * cosf(M_PI * (time - (tRec - tFadeOut)) / tFadeOut + M_PI);
+    } else {
+        window = 0.0;
+    }
+
+    if(time <= 0.0f) {
+        waveformFreqRads = wMin;
+        output = 0.0f;
+    } else if (time <= tConst) {
+        waveformFreqRads = wMin;
+        output = window * magnitude_scale * magnitude * sinf(wMin * time - wMin * tConst);
+    } else if (time <= tRec) {
+        waveformFreqRads = wMin * expf(B * (time - tConst) / (tRec - tConst));
+        output = window * magnitude_scale * magnitude * sinf((wMin * (tRec - tConst) / B) * (expf(B * (time - tConst) / (tRec - tConst)) - 1));
```

Comment:
Do we need to check that tRec and tConst are not the same which (if they were) could result in a divide by zero?

Reply:
Nice catch. Done.